### PR TITLE
Enable monitoring buffer for elastic-agent

### PR DESCRIPTION
**Pull request by michel-laterman**
## What does this PR do?

Enable the monitoring buffer for the elastic-agent and beats that it starts.
Add metrics collection (from buffers) into the `diagnostics collect` bundle if enabled.

## Why is it important?

Allows us to gather metrics data for debugging if agent->ES communications are failing.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation [PR](https://github.com/elastic/observability-docs/pull/1627)
- [x] I have made corresponding change to the default configuration files
- [] ~~I have added tests that prove my fix is effective or that my feature works~~
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

Start agent and run `elastic-agent diagnostics collect`. Metrics should not be included in the archive by default.
Enable metrics buffers by adding `agent.monitoring.http.buffer.enabled: true` then rerun `elastic-agent diagnostics collect`, metrics (json) files should be added to the archive

## Related issues

- Relates #29776

### CODE REVIEW

File: x-pack/elastic-agent/_meta/config/common.reference.p2.yml.tmpl [hunk 1]
```diff
@@ -119,6 +119,12 @@ inputs:
 #       host: localhost
 #       # Port on which the HTTP endpoint will bind. Default is 0 meaning feature is disabled.
 #       port: 6791
+#       # Metrics buffer config
+#       buffer:
+#           enabled: false
+#           period: 10s
+#           size: 60
+#           namespaces: ['stats']
```
Comment (michel-laterman) [inline x-pack/elastic-agent/_meta/config/common.reference.p2.yml.tmpl:127]: I've added the full buffer config here, but I'm not sure if we should. Currently the `enabled` flag is the only one that's passed (it's injected via command line args at the moment).

File: x-pack/elastic-agent/CHANGELOG.next.asciidoc [hunk 2]
```diff
@@ -161,4 +161,5 @@
 - Discover changes in Kubernetes nodes metadata as soon as they happen. {pull}23139[23139]
 - Add results of inspect output command into archive produced by diagnostics collect. {pull}29902[29902]
 - Add support for loading input configuration from external configuration files in standalone mode. You can load inputs from YAML configuration files under the folder `{path.config}/inputs.d`. {pull}30087[30087]
+- Add support for enabling the metrics buffer endpoint in the elastic-agent and beats it runs. diagnostics collect command will gather metrics-buffer data if endabled. {pull}30471[30471]
```
Comment (lykkin) [inline x-pack/elastic-agent/CHANGELOG.next.asciidoc:164]:
```suggestion
- Add support for enabling the metrics buffer endpoint in the elastic-agent and beats it runs. diagnostics collect command will gather metrics-buffer data if enabled. {pull}30471[30471]
```

File: x-pack/elastic-agent/pkg/agent/cmd/diagnostics.go [hunk 3]
```diff
@@ -204,6 +204,15 @@ func diagnosticsCollectCmd(streams *cli.IOStreams, fileName, outputFormat string
 		return fmt.Errorf("failed to communicate with Elastic Agent daemon: %w", err)
 	}
 
+	metrics, err := gatherMetrics(innerCtx)
+	if err == context.DeadlineExceeded {
+		return errors.New("timed out after 30 seconds trying to connect to Elastic Agent daemon")
+	} else if err == context.Canceled {
+		return nil
+	} else if err != nil {
+		return fmt.Errorf("failed to communicate with Elastic Agent daemon: %w", err)
```
Comment (lykkin) [inline x-pack/elastic-agent/pkg/agent/cmd/diagnostics.go:213]: would outputting a partial response would be useful when metrics error? e.g. the metrics can't be harvested, but the config info can be.

**Review by lykkin — APPROVED** [review]
LGTM, just some questions for context:

beat metrics are collected in the diag command over grpc by way of an http endpoint. does that sound right?

are there any docs i can read up on the metrics libbeat exports this way?

File: x-pack/elastic-agent/pkg/agent/cmd/diagnostics.go (hunk 3 above)
Comment (michel-laterman) [reply x-pack/elastic-agent/pkg/agent/cmd/diagnostics.go:213]: yeah, that's part of another task: https://github.com/elastic/beats/issues/29818

**Review by ph — CHANGES_REQUESTED** [review]
@michel-laterman Can we add a few tests to ensure the behavior?

File: x-pack/elastic-agent/_meta/config/common.reference.p2.yml.tmpl (hunk 1 above)
Comment (blakerouse) [reply x-pack/elastic-agent/_meta/config/common.reference.p2.yml.tmpl:127]: I agree if its not exposed then it should be hidden.

**Review by blakerouse — APPROVED** [review]
Looks good to me, minus the one comment.

**Comment by michel-laterman** [comment]
@ph, as we discussed I think testing for the diagnostics commands should be added to e2e tests, i've made an issue to track this here: https://github.com/elastic/e2e-testing/issues/2204

**Review by ph — APPROVED** [review]
LGTM

**Comment by jlind23** [comment]
@michel-laterman will you be able to merge it before your time off?

**Comment by michel-laterman** [comment]
yep, as soon as the e2e passes i'll merge

**Comment by michel-laterman** [comment]
I'm going to force this through, all checks succeeded in https://github.com/elastic/beats/pull/30471/commits/c8339ad20f15e8c237e2617643f999c85ad4552b and I have only changed default config to remove unused values.

**Pull request merged**